Добавлена валидация при приеме запросов

Desann69 opened the pull request.

Desann69 committed:
Добавлена валидация при приеме запросов
Порт 5433 по умолчанию

IlyaKo commented on src/backend/LosTomates.PetHolidays.WebApi/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -22,13 +23,12 @@ internal static IServiceCollection AddSwagger(this IServiceCollection services)
     internal static IServiceCollection AddDatabaseContext(this IServiceCollection services, IConfiguration configuration)
     {
-        var connectionString = configuration.GetConnectionString("CoreDb");
+        string? connectionString = configuration.GetConnectionString("CoreDb");
```
Здесь и дальше по коду почему-то изменен только `var` на конкретный тип. 
Я бы предложил в рамках этого PR не менять ну и вообще договориться использовать `var` где это возможно - так как это очень удобно не отвлекать внимание на конкретный тип. Если что - IDE умеют подсказывать что это за тип очень хорошо, а во многих случаях вообще не обязательно знать что это за тип.

Desann69 replied:
Насчет коррекций в PR согласен, но лучше использовать, мне кажется, явные типы, чтобы быстрее улавливать контекст

IlyaKo replied:
Если согласен насчет того, чтобы не менять в PR - тогда надо бы вернуть `var` на место 😃

IlyaKo commented on src/backend/LosTomates.PetHolidays.WebApi/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -37,4 +37,11 @@ internal static IServiceCollection AddDatabaseContext(this IServiceCollection se
+
+    internal static IServiceCollection AddFluentValidation(this IServiceCollection services)
+    {
+        //services.AddValidatorsFromAssembly(Assembly.GetExecutingAssembly());
```
Лучше не допускать закомментированного кода в основные ветки проекта - т.е. где-то для тестирования, в отдельной ветке - это может быть удобно, но в целом лучше убирать из кода.

IlyaKo commented on src/backend/LosTomates.PetHolidays.WebApi/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -37,4 +37,11 @@ internal static IServiceCollection AddDatabaseContext(this IServiceCollection se
+    internal static IServiceCollection AddFluentValidation(this IServiceCollection services)
+    {
+        //services.AddValidatorsFromAssembly(Assembly.GetExecutingAssembly());
+        services.AddSingleton<IValidator<HotelEditDto>,  HotelEditDtoValidator>();
```
Предлагаю здесь находить все валидаторы автоматически, так как ты выше закомментировал - `services.AddValidatorsFromAssembly`

IlyaKo commented on src/backend/LosTomates.PetHolidays.WebApi/Validations/HotelEditDtoValidator.cs:
```diff
@@ -0,0 +1,30 @@
+using LosTomates.PetHolidays.Application.Hotels;
+using LosTomates.PetHolidays.DataAccess;
+
+public class HotelEditDtoValidator : AbstractValidator<HotelEditDto>
```
Предлагаю весь код перенести в проект Application, там есть уже папка Hotels для этих штук

IlyaKo commented on src/backend/LosTomates.PetHolidays.WebApi/Endpoints/HotelEndpoints.cs:
```diff
@@ -28,20 +33,37 @@ public static void Map(WebApplication app)
-                .Produces<HotelView>(StatusCodes.Status200OK);
+        mapGroup.MapPost(string.Empty, async (IHotelService service, IValidator<HotelEditDto> validateService, HotelEditDto dto) =>
+        {
+            //IValidator<HotelEditDto> validateService = app.Services.GetService<IValidator<HotelEditDto>>();
```
Весь код по валидации лучше перенести в сервисный класс `HotelService` - чтобы этот класс отвечал именно за эндпоинты и их маппинг

IlyaKo commented on src/backend/LosTomates.PetHolidays.WebApi/Endpoints/HotelEndpoints.cs:
```diff
@@ -28,20 +33,37 @@ public static void Map(WebApplication app)
-            => await service.Update(hotelId, dto))
+        mapGroup.MapPut("{hotelId:int}", async (IHotelService service, IValidator<HotelEditDto> validateService, int hotelId, HotelEditDto dto) => 
+        {
+            FluentValidation.Results.ValidationResult validationResult = validateService.Validate(dto);
```
Есть метод `ValidateAndThrow` или как-то так - чтобы не писать каждый раз однотипный код

Desann69 commented on .DS_Store:
Принято! Добавлю со следующим PR

IlyaKo replied:
На то комментарии и есть, чтобы в рамках этого PR исправления вносить - так легче будет отслеживать комментарии

IlyaKo replied:
Что-то файлики не исчезли, может надо еще руками их удалить? Попробуй прямо в браузере эти файлики удалить:
![image](https://github.com/user-attachments/assets/473e9998-f558-4824-aa4b-4eac5c770f7b)

IlyaKo replied:
Удалил в итоге файлики сам

Desann69 committed:
Доработка валидации

IlyaKo committed:
updated NuGet packages

IlyaKo commented:
Можно подтянуть еще изменения из ветки Develop и добавить Middleware обработчик для исключений валидации, так как сейчас возвращается ошибка 500

IlyaKo commented:
@Desann69 Можно опубликовать обратно PR, а то он всё еще в draft'e

Desann69 committed:
Корректировка валидации

Desann69 committed:
Исправлены правила для валидации

Desann69 committed:
Вернул var

Desann69 committed:
Merge branch 'develop' into Tasks/12-fluent-validation-webapi

Desann69 committed:
Ручная склейка

IlyaKo commented on src/backend/LosTomates.PetHolidays.WebApi/Program.cs:
```diff
@@ -59,8 +60,8 @@ private static void ConfigurePipeline(WebApplication app)
-        using var scope = services.CreateScope();
-        var seedService = scope.ServiceProvider.GetRequiredService<SeedService>();
+        using IServiceScope scope = services.CreateScope();
+        SeedService seedService = scope.ServiceProvider.GetRequiredService<SeedService>();
```
Тут тоже `var` вернуть как было

IlyaKo commented on src/backend/LosTomates.PetHolidays.Application/Hotels/HotelService.cs:
```diff
@@ -61,5 +80,11 @@ public async Task Delete(int entityId)
-        => await dbContext.Hotels.FirstOrDefaultAsync(x => x.Id == entityId);
-}
+    {     
+        bool isValid = entityId >= 0; 
```
Это private метод, тут не должно быть такого исключения

IlyaKo commented on src/backend/LosTomates.PetHolidays.Application/Exceptions/ValidationBadRequest.cs:
Я бы вообще удалил этот класс, если нужно бросить исключение валидации - тогда я бы использовал то, которое уже есть - `FluentValidation.ValidationException`

Desann69 committed:
Убрано из цепочки валидации When  выражение

Desann69 reviewed (commented):
When убрал
Но с DS_Store убрать из первого коммита не удалось

Desann69 committed:
Удалил CustomValidateEx
Вернул var
Удалена избыточная валидация
Удалаена перегрузка метода обработки валидации

IlyaKo committed:
removed Mac files

IlyaKo commented on src/backend/LosTomates.PetHolidays.Application/Hotels/HotelService.cs:
```diff
@@ -10,27 +11,38 @@ public sealed class HotelService : IHotelService
 
     public async Task<HotelView?> GetById(int entityId)
     {
+        bool isValid = entityId >= 0; 
```
В delete тоже надо убрать эту валидацию

Desann69 committed:
Убрана валидация id в delete
